### ISSUE TITLE
Move, rotate and scale Point Clouds

### ISSUE THREAD

**Issue by Optimus13** [SFraissTU/BA_PointCloud#8]
Hello!

First of all, I want to say thank you for sharing this project, it's simply awesome :)

I'm working on a point cloud visualizer in VR with unity and I want to add translation, rotation and scaling of the point clouds.

I've tried parenting all the gameObjects created by the DefaultMeshConfiguration and move the parent but I think the boudingBoxes aren't moving so another point cloud is rendered.
Do you have any method to handle these behaviors? If not, do you maybe have some clues to give me so I can try doing it myself?

Thanks a lot for your answer!

**Comment by SFraissTU** [comment]
Hi! Sorry for my late response!

Translation is currently supported. You just need to set the position of the transform component of the Point Cloud Set to the desired position. However, this position needs to be specified before the rendering starts. Changing the transform during runtime will not affect the point cloud. Also, scale and rotation are ignored so far.
You are right that when you change the transform of the created GameObjects, the bounding boxes are not updated. One trick to overcome this would be to have an additional camera, that always looks at the point cloud, as if no transformation had happened, while also moving along with the real camera. 
So for example, if your camera is at (2,0,0) and the point cloud is not transformed, this new camera should also be at (2,0,0). If however, you translate the point cloud by +3 along the y-axis, the new camera is now at (2,-3,0) (so if the point cloud had not been transformed, it would see the same as the real camera sees now with the transformed point cloud, because it's y-difference is 3 both times). And when you now move the real camera to (4,0,0), the new camera switches to (4,-3,0).
So basically you always apply the inverse transformation of the point cloud to this new camera.
You would need to implement this yourself. This camera you can then specify as the User Camera in the Point Cloud Set.
I hope this makes sense. I didn't try it myself, but I imagine it should work.

Another way would be to dig a bit into the code and take the actual Node-objects and transform the bounding boxes specified there. The root nodes are stored in the V2Renderer during initialization. You can traverse the node hierarchy and adapt the bounding boxes yourself, but you will need to adapt the code quite a bit.

Cheers, Simon

**Comment by nabil-dbz** [comment]
Hi Simon,

I, first, would like to thank you for sharing your interesting project with us. It's really a good and efficient way of rendering large point cloud models in unity. 

I am working on a VR project and I would like to add an experience that includes some point cloud models. For that, I would like to ask you two questions:
1- Have you tried your technique on VR ?
2- Is there a way to change the position of some points (nodes) without transforming the root point when the game's running of course? For example, if they collide with another collider? 

Thank you in advance and good luck!

Best,
Nabil

**Comment by SFraissTU** [comment]
Hi. Thank you for your feedback!

1. No I didn't have the chance yet, I might do so in autumn, as then I will have a headset available. But I've heard from other people who successfully managed to use it for VR.
2. Currently not. In my last answer in this thread I described my idea how you could implement dynamic transformations of the whole point cloud yourself, and I hope I will implement this myself as well in the next months. Transforming just a few points or nodes however is even more tricky, as this would change the bounding box hierarchy, which would then be needed to rebuild. I guess it should be possible if you use a StaticPointCloudSet (which loads the whole cloud at once and does not take your camera position into consideration for what to load - but this is not usable for very big clouds). Then you could just transform the created GameObjects for each Node and your done with it. But for DynamicPointCloudSet, this seems like a lot of work, and I do not plan on implementing this currently.

Simon

**Comment by nabil-dbz** [comment]
Thank you Simon for your response. I actually have been trying to do that with the StaticPointCloudSet just to have an idea on how it might me done using DynamicPointCloudSet. The problem is that I only managed to transform the whole Point Cloud Set as a game object, but what I want is to transform some points only (only a part of the game object).

**Comment by SFraissTU** [comment]
Both Static and DynamicPointCloudSets create a bunch of GameObjects in the scene. Basically, each node of the octree structure is represented as a GameObject attached with the a Mesh Filter and a Mesh Renderer. The name of the GameObject contains the address of the node inside the octree (for example "r" is the root node, "r02" is the third child of the first child of the root node. By changing the transform of these objects, you can move parts of the point cloud around. See this screenshot: http://sryus.atria.uberspace.de/gos.png

However, you will probably run into a problem. The way the Potree octree format is designed, the points of a certain area are distributed over all levels of the hierarchy. So if you would want to transform exactly a certain area of the point cloud, it wouldn't be enough to just transform the node, whose bounding box corresponds to the desired area and all of it's children. Points of this area are also contained in all parent nodes. However, the parents also contain points from other regions. So really you would ultimately need to change the vertexData of the meshes of the parents to get a correct result.

This whole thing is not really designed to modify point clouds. The used format is really useful for displaying not-changing point clouds (because you can display a rough preview of the point cloud by just loading the higher levels), but when you start modifying them, things will probably get really tricky.

**Comment by akharroubi** [comment]
![Capture d’écran (20)](https://user-images.githubusercontent.com/41165081/60726777-a95fe900-9f3c-11e9-9dfb-117580820083.png)

**Comment by nabil-dbz** [comment]
Hi Simon, 

Thank you for your response. I am sorry for the delay! Yeah, I just realized that and I think that I should look for another technique to display the point cloud without the Potree octree format. Do you have any suggestions or points I could start from ?

**Comment by SFraissTU** [comment]
Hi! No, I'm currently not familiar with anything I could suggest you. I imagine it's tough to combine your editing requirements with the ability to dynamically load a big point cloud depending on the camera position. Because you would also need to store your changes on the disk, otherwise they would be gone if you move away from the edited part and revisit it later.
If you only use point clouds that are sufficiently small it would be easier to just load the whole point cloud into memory, maybe using an octree structure where all points are stored in the leaves.
Simon

**Comment by SFraissTU** [comment]
Hello again! Back to your original question @Optimus13 .
It's been a while, but now it's possible to transform point cloud sets using translation, rotation and scale. This is also possible during runtime. The way I implemented this was to set the created node game objects as children to the point cloud sets. Additionally, in the V2Renderer, I calculate the inverse transformation and apply this one to the camera, such that the viewing queries still work!
Cheers!
